### PR TITLE
CI: fix CI to run consistent tests 100% of the time

### LINKED ISSUE #32939: Make CI jobs pass 100/100 times
We've got a new weekly job that evaluates the reliability of our CI by running it several times. Here's the latest run: https://github.com/commaai/openpilot/actions/runs/9837241808

Goal is to get all tests to pass 100/100 times.

### PR DESCRIPTION
**Pull request by bongbui321**
~Recently, I only see that mostly unit_test section is failing, and from looking at the failing logs, it seems that there is this `Fatal Python error: Bus error` that causes crashes in pytest-xdist workers~. ~This error is most likly from workers trying to access a unavailable memory region, so I just increase the shm-size of docker from 1GB -> 2GB~. 

~From running 2 runs of this new size (50 runs of unit_test each), I don't see that error happening anymore. `~
~run 1: https://github.com/commaai/openpilot/actions/runs/10115487281/job/27976375147~
~run 2: https://github.com/commaai/openpilot/actions/runs/10115625540/job/27976846602~

~The failing tests of these two runs are from a known flaky test in `test_loggerd.py` and I guess it has some weird association with matrix that causing it to fail (didn't investigate further)~, ~but should be fine if we have one run of `unit_test`~

~I did another run to make sure that if the shm-size is 1GB then we would see the bus error in unittest again~

~https://github.com/commaai/openpilot/actions/runs/10115830453/job/27977440198~

~If the issues still continues after this is merged, I'll happy to finish it.~

This is probably something due to pytest-xdist running tests that mess around with threading.

resolves [#32939](https://github.com/commaai/openpilot/issues/32939)
- [x] serial testing. Works!

### CODE REVIEW

**Comment by bongbui321** [comment]
wasn't the issue I initially thought it was.

File: selfdrive/ui/tests/test_translations.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ def test_missing_translation_files(self):
     assert os.path.exists(os.path.join(TRANSLATIONS_DIR, f"{self.file}.ts")), \
                     f"{self.name} has no XML translation file, run selfdrive/ui/update_translations.py"
 
+  @pytest.mark.serial
```
Comment (sshane) [inline selfdrive/ui/tests/test_translations.py:37]: Why doesn't `translations_dir` work to solve this? Is `TemporaryDirectory` somehow returning the same directory for two workers?

File: .github/workflows/selfdrive_tests.yaml [hunk 2]
```diff
@@ -25,7 +25,7 @@ env:
...
+  RUN: docker run --shm-size 2G -v $PWD:/tmp/openpilot -w /tmp/openpilot -e CI=1 -e PRE_COMMIT_HOME=/tmp/pre-commit -e PYTHONWARNINGS=error -e FILEREADER_CACHE=1 -e PYTHONPATH=/tmp/openpilot -e NUM_JOBS -e JOB_ID -e GITHUB_ACTION -e GITHUB_REF -e GITHUB_HEAD_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_RUN_ID -v $GITHUB_WORKSPACE/.ci_cache/pre-commit:/tmp/pre-commit -v $GITHUB_WORKSPACE/.ci_cache/scons_cache:/tmp/scons_cache -v $GITHUB_WORKSPACE/.ci_cache/comma_download_cache:/tmp/comma_download_cache -v $GITHUB_WORKSPACE/.ci_cache/openpilot_cache:/tmp/openpilot_cache $BASE_IMAGE /bin/bash -c
```
Comment (sshane) [inline .github/workflows/selfdrive_tests.yaml]: Any way we can check the runner logs to see if it ran out of memory or not? It does seem probable, wonder what has increased in memory usage for the last few months though. From adding more tests to this one job, or a memory leak?

Comment (sshane) [reply .github/workflows/selfdrive_tests.yaml]: test_car_interfacees.py seems to increase my machine's memory usage by 1GB with 8 jobs!

File: selfdrive/ui/tests/test_translations.py (hunk 1 above)
Comment (bongbui321) [reply selfdrive/ui/tests/test_translations.py]: This was removed since I used -n0 for the test_translation.py that make it serial.

**Comment by bongbui321** [comment]
Current attempt to serialize tests that seems to fail tests running in parallel with pytest-xdist.

https://github.com/pytest-dev/pytest/issues/3216#issuecomment-742143197

File: .github/workflows/selfdrive_tests.yaml (hunk 2 above)
Comment (sshane) [reply .github/workflows/selfdrive_tests.yaml]: `pytest-memray`, not sure how accurate this is:

```
Allocation results for selfdrive/car/tests/test_car_interfaces.py::TestCarInterfaces::test_car_interfaces_007_BUICK_REGAL at the high watermark

         📦 Total memory allocated: 39.1MiB
         📏 Total allocations: 390
         📊 Histogram of allocation sizes: | ▄█  |
         🥇 Biggest allocating functions:
                - get_kalman_gain:/home/batman/openpilot/common/simple_kalman.py:7 -> 32.0MiB
                - getstate:/home/batman/.pyenv/versions/3.11.4/lib/python3.11/random.py:168 -> 1.0MiB
                - <dictcomp>:/home/batman/openpilot/selfdrive/test/fuzzy_generation.py:50 -> 1.0MiB
                - __init__:/home/batman/openpilot/selfdrive/car/gm/carstate.py:18 -> 1.0MiB
                - getstate:/home/batman/.pyenv/versions/3.11.4/lib/python3.11/random.py:168 -> 1.0MiB
```

Comment (bongbui321) [reply .github/workflows/selfdrive_tests.yaml]: after a bit of testing and research, I'm not sure if this is an issue of memory usage. 

https://github.com/pytest-dev/pytest/issues/3216#issuecomment-742143197. After serializing some tests, everything passes pretty consistently

Comment (sshane) [reply .github/workflows/selfdrive_tests.yaml]: You can monitor the usage for a run to see how close it is to 1GB currently, that'll give you a good idea if it contributes.

Comment (bongbui321) [reply .github/workflows/selfdrive_tests.yaml]: other pytest monitoring tools seems to only measure singular test only which makes it hard to measure memory usage. I manually look at `htop` and see that running a `pytest -n8 'no slow'` takes > 2.5GB

**Comment by bongbui321** [comment]
findings:
- Increase the shm-size seems to not have the python bus error anymore but it still has some random test failures for ones that have multiprocessing
- After researching for a while, a lot of people seems to have this issue, and it seems to only happen on CI ubuntu machines within a container
- Seems nearly impossible to root-cause the problem since it is not easily reproducible, too random.
- Serializing tests that involves multiple processes seems to work

**Comment by adeebshihadeh** [comment]
> * Increase the shm-size seems to not have the python bus error anymore but it still has some random test failures for ones that have multiprocessing

Are the errors now specific to a particular test? If so, that's worth including here. 

> * Serializing tests that involves multiple processes seems to work

It still seems like we don't know what's going on here. What if it's this https://github.com/commaai/openpilot/issues/27512?

**Comment by bongbui321** [comment]
> Are the errors now specific to a particular test? If so, that's worth including here.

Included in comments.
- `test_loggerd.py::test_rotation` runs multiple processes
- `test_translation.py` has some I/O operations which might be the cause of some failing tests. 
A lot of people experience these random worker crashes using pytest-xdist, and there haven't been to have good repro or clear solution ([comment](https://github.com/coiled/benchmarks/issues/404#issuecomment-1270109736))

> It still seems like we don't know what's going on here. What if it's this https://github.com/commaai/openpilot/issues/27512?

Did a quick run`unit_tests` 50 times with deanlee's latest temporary fix and still experiencing these failures ([run](https://github.com/bongbui321/openpilot/actions/runs/10152883757)). This is likely to be a bug within pytest-xdist

**Comment by adeebshihadeh** [comment]
I looked into this briefly, and seems like there's multiple things going on and the issues are interacting with each other. When that happens, it's important to carefully fix each one.
* worker crash -> OOM due to shm size
* `test_rotation` failure -> encoderd is crashing

This seems quite reproducible, so we should be able to make progress on it.

I pushed the shm bump here: https://github.com/commaai/openpilot/commit/76fd5b00f1f169013334e06031e6b063c16cf5a5

encoderd crash from https://github.com/commaai/openpilot/actions/runs/10174071612/job/28139196053:
```
>       assert len(diff) == 0, f"didn't get all expected files. run={_} seg={n} {route_path=}, {diff=}\n{logged=} {expected_files=}"
E       AssertionError: didn't get all expected files. run=(1928, 1208, 4804608, 2048, 2490368) seg=0 route_path='/root/.commadd14e5bfc945470/media/0/realdata/00000000--7bf74a8990', diff={'ecamera.hevc', 'fcamera.hevc', 'dcamera.hevc', 'qcamera.ts'}
E         logged={'qlog', 'rlog'} expected_files={'ecamera.hevc', 'qlog', 'fcamera.hevc', 'rlog', 'dcamera.hevc', 'qcamera.ts'}
E       assert 4 == 0
E        +  where 4 = len({'dcamera.hevc', 'ecamera.hevc', 'fcamera.hevc', 'qcamera.ts'})

system/loggerd/tests/test_loggerd.py:185: AssertionError
----------------------------- Captured stdout call -----------------------------
Starting listener for: camerad
system/loggerd/loggerd.cc: logging to /root/.comma81d1a5fdea744a9/media/0/realdata/00000000--e3cf4e4207--0
system/loggerd/loggerd.cc: closing logger
----------------------------- Captured stderr call -----------------------------
encoderd: msgq_repo/msgq/msgq.cc:232: int msgq_msg_send(msgq_msg_t *, msgq_queue_t *): Assertion `3 * total_msg_size <= q->size' failed.
----------------------------- Captured stdout call -----------------------------
Starting listener for: camerad
system/loggerd/loggerd.cc: logging to /root/.comma285a06b46ef64a8/media/0/realdata/00000000--6894b06f08--0
system/loggerd/loggerd.cc: closing logger
----------------------------- Captured stderr call -----------------------------
encoderd: msgq_repo/msgq/msgq.cc:232: int msgq_msg_send(msgq_msg_t *, msgq_queue_t *): Assertion `3 * total_msg_size <= q->size' failed.
----------------------------- Captured stdout call -----------------------------
Starting listener for: camerad
system/loggerd/loggerd.cc: logging to /root/.commadd14e5bfc945470/media/0/realdata/00000000--7bf74a8990--0
system/loggerd/loggerd.cc: closing logger
----------------------------- Captured stderr call -----------------------------
encoderd: msgq_repo/msgq/msgq.cc:232: int msgq_msg_send(msgq_msg_t *, msgq_queue_t *): Assertion `3 * total_msg_size <= q->size' failed.
```

**Pull request closed without merging**